common/flatpak-dir.c

```diff
@@ -7502,6 +7505,179 @@ flatpak_dir_create_system_child_repo (FlatpakDir   *self,
+                                       NULL);
+
+  g_autoptr(GFile) appstream_file = g_file_new_for_path (appstream_path);
+  as_store_from_file (store, appstream_file, NULL, cancellable, &local_error);
+  success = (local_error == NULL);
+
+  /* We want to ignore ENOENT error as it is harmless and valid
+   * FIXME: appstream-glib doesn't have granular file-not-found error
+   * See: https://github.com/hughsie/appstream-glib/pull/268 */
+  if (local_error != NULL &&
+      g_str_has_suffix (local_error->message, "No such file or directory"))
+    g_clear_error (&local_error);
+  else if (local_error != NULL)
+    g_propagate_error (error, g_steal_pointer (&local_error));
+
+  return success;
+}
+
+/* Convert an #EpcAppFilterOarsValue to an #AsContentRatingValue. This is
+ * actually a trivial cast, since the types are defined the same; but throw in
+ * a static assertion to be sure. */
+static AsContentRatingValue
+convert_app_filter_oars_value (EpcAppFilterOarsValue filter_value)
+{
+  G_STATIC_ASSERT (AS_CONTENT_RATING_VALUE_LAST == EPC_APP_FILTER_OARS_VALUE_INTENSE + 1);
+
+  return (AsContentRatingValue) filter_value;
+}
+
+/* Get the AppStream data for @ref, loading it from the locally cached AppStream
+ * XML. If the AppStream file doesn’t exist, or if @ref_string is not listed in
+ * it, %NULL will be returned without an error. An error will be returned if
+ * parsing the AppStream file fails.
+ *
+ * If @arch is %NULL, the default architecture will be used.
+ *
+ * If the @ref_string is for a runtime, %NULL will be returned immediately with
+ * no error.
+ *
+ * A strong reference is returned on the #AsApp.
+ */
+static AsApp *
+get_appstream_for_ref (FlatpakDir          *self,
+                       FlatpakRemoteState  *state,
+                       const char          *ref_string,
+                       const char          *arch,
+                       GCancellable        *cancellable,
+                       GError             **error)
+{
+  g_autoptr(FlatpakRef) ref = NULL;
+
+  ref = flatpak_ref_parse (ref_string, error);
+  if (ref == NULL)
+    return NULL;
+
+  /* Runtimes don’t have appstream. */
+  if (flatpak_ref_get_kind (ref) != FLATPAK_REF_KIND_APP)
+    return NULL;
+
+  g_autoptr(AsStore) store = as_store_new ();
+  g_autofree char *unique_id = NULL;
+
+  if (!flatpak_dir_load_appstream_store (self, state->remote_name, arch, store,
+                                         cancellable, error))
+    return NULL;
+
+  unique_id = as_utils_unique_id_build (AS_APP_SCOPE_UNKNOWN,
+                                        AS_BUNDLE_KIND_FLATPAK,
+                                        NULL,  /* origin */
+                                        AS_APP_KIND_DESKTOP,
+                                        flatpak_ref_get_name (ref),
+                                        flatpak_ref_get_branch (ref));
+  AsApp *app = as_store_get_app_by_unique_id (store, unique_id,
```

Comment:
If the same app is available for 2 different architectures, what gets returned here? For instance, you can install and run i386 flatpaks on an x86_64 system. Can a user circumvent parental controls by installing the i386 one?

Reply:
I believe flatpak generates the `appstream.xml` file for the repository from the app’s OSTree commit, and it injects the `<bundle>` elements into it. Therefore the appstream is always generated separately for each architecture of an app, and so should always contain the right OARS data. See [`extract_appstream()` in `flatpak-utils.c`](https://github.com/flatpak/flatpak/blob/bf7af547aa65eeee0f9c7a4a73e2c88eb4b84693/common/flatpak-utils.c#L3390).

Reply:
Is the arch name included in the blacklist entry?

Reply:
Blacklist entry? The `appstream.xml` file is stored per-arch in `/var/lib/flatpak/appstream/${remote_name}/${arch}/appstream.xml{,.gz}`.

Reply:
My concern was that you could circumvent the filter by installing an app from a foreign arch (i386 on x86_64) but I now see that the appstream DB that gets loaded is based on the arch being run.